clippy_utils/src/lib.rs

```diff
@@ -2884,3 +2885,111 @@ op_utils! {
     Shl    ShlAssign
     Shr    ShrAssign
 }
+
+/// Produces a path from a local caller to the type of the called method. Suitable for user
+/// output/suggestions.
+///
+/// Returned path can be either absolute (for methods defined non-locally), or relative (for local
+/// methods).
+pub fn get_ufcs_type_name<'tcx>(
+    tcx: TyCtxt<'tcx>,
+    from: LocalDefId,
+    method: DefId,
+    args: GenericArgsRef<'tcx>,
+) -> String {
+    let assoc_item = tcx.associated_item(method);
+    let def_id = assoc_item.container_id(tcx);
+    match assoc_item.container {
+        rustc_ty::TraitContainer => get_path_to_callee(tcx, from, def_id),
+        rustc_ty::ImplContainer => {
+            let ty = tcx.type_of(def_id).instantiate_identity();
+            match ty.kind() {
+                rustc_ty::Adt(adt, _) => get_path_to_callee(tcx, from, adt.did()),
+                rustc_ty::Array(..)
+                | rustc_ty::Dynamic(..)
+                | rustc_ty::Never
+                | rustc_ty::RawPtr(_)
+                | rustc_ty::Ref(..)
+                | rustc_ty::Slice(_)
+                | rustc_ty::Tuple(_) => format!("<{}>", EarlyBinder::bind(ty).instantiate(tcx, args)),
```

Comment:
We need to handle these too, I'm pretty sure. This should recursively go through each nested type and get the relative path for each.

Not an issue for the lint in question, but if somebody uses this on something like `(super::A, super::super::B)` it'll still be broken afaik.